core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -322,51 +327,81 @@ public void close()
+                DataSize targetPartitionSourceSize)
         {
             checkArgument(bucketNodeMap.isPresent() || fragment.getPartitionedSources().isEmpty(), "bucketNodeMap is expected to be set when the fragment reads partitioned sources (tables)");
             Map<PlanNodeId, SplitSource> splitSources = splitSourceFactory.createSplitSources(session, fragment);
+
             return new HashDistributionTaskSource(
                     splitSources,
+                    getExchangeForHandleMap(sourceExchanges, exchangeSourceHandles),
                     getPartitionedExchangeSourceHandles(fragment, exchangeSourceHandles),
                     getReplicatedExchangeSourceHandles(fragment, exchangeSourceHandles),
                     splitBatchSize,
                     getSplitTimeRecorder,
                     bucketToPartitionMap,
                     bucketNodeMap,
-                    fragment.getPartitioning().getConnectorId());
+                    fragment.getPartitioning().getConnectorId(),
+                    targetPartitionSplitWeight, targetPartitionSourceSize);
+        }
+
+        private static IdentityHashMap<ExchangeSourceHandle, Exchange> getExchangeForHandleMap(
+                Map<PlanFragmentId, Exchange> sourceExchanges,
+                Multimap<PlanFragmentId, ExchangeSourceHandle> exchangeSourceHandles)
+        {
+            IdentityHashMap<ExchangeSourceHandle, Exchange> exchangeForHandle = new IdentityHashMap<>();
+            for (Map.Entry<PlanFragmentId, ExchangeSourceHandle> entry : exchangeSourceHandles.entries()) {
+                PlanFragmentId fragmentId = entry.getKey();
+                ExchangeSourceHandle handle = entry.getValue();
+                Exchange exchange = sourceExchanges.get(fragmentId);
+                requireNonNull(exchange, "Exchange not found for fragment " + fragmentId);
+                exchangeForHandle.put(handle, exchange);
+            }
+            return exchangeForHandle;
         }
 
         public HashDistributionTaskSource(
                 Map<PlanNodeId, SplitSource> splitSources,
+                IdentityHashMap<ExchangeSourceHandle, Exchange> exchangeForHandle,
                 Multimap<PlanNodeId, ExchangeSourceHandle> partitionedExchangeSourceHandles,
                 Multimap<PlanNodeId, ExchangeSourceHandle> replicatedExchangeSourceHandles,
                 int splitBatchSize,
                 LongConsumer getSplitTimeRecorder,
                 int[] bucketToPartitionMap,
                 Optional<BucketNodeMap> bucketNodeMap,
-                Optional<CatalogName> catalogRequirement)
+                Optional<CatalogName> catalogRequirement,
+                long targetPartitionSplitWeight,
+                DataSize targetPartitionSourceSize)
         {
             this.splitSources = ImmutableMap.copyOf(requireNonNull(splitSources, "splitSources is null"));
+            this.exchangeForHandle = new IdentityHashMap<>();
```

Comment:
nit: I wonder if it is necessary to create a defensive copy, given this is a private constructor?

Reply:
It is negligible. I will keep it